extensions/libarchive/ManipulateArchive.cpp

```diff
@@ -0,0 +1,309 @@
+namespace nifi {
+namespace minifi {
+namespace processors {
+
+core::Property ManipulateArchive::Operation("Operation", "Operation to perform on the archive", "");
+core::Property ManipulateArchive::Target("Target", "The path within the archive to perform the operation on", "");
+core::Property ManipulateArchive::Destination("Destination", "Destination for operations (move or copy) which result in new entries", "");
+core::Property ManipulateArchive::Before("Before", "For operations which result in new entries, places the new entry before the entry specified by this property", "");
+core::Property ManipulateArchive::After("After", "For operations which result in new entries, places the new entry after the entry specified by this property", "");
+core::Relationship ManipulateArchive::Success("success", "success operational on the flow record");
+
+char const* ManipulateArchive::OPERATION_REMOVE = "remove";
+char const* ManipulateArchive::OPERATION_COPY =   "copy";
+char const* ManipulateArchive::OPERATION_MOVE =   "move";
+char const* ManipulateArchive::OPERATION_TOUCH =  "touch";
+
+void ManipulateArchive::initialize() {
+    //! Set the supported properties
+    std::set<core::Property> properties;
+    properties.insert(Operation);
+    properties.insert(Target);
+    properties.insert(Destination);
+    properties.insert(Before);
+    properties.insert(After);
+    setSupportedProperties(properties);
+
+    //! Set the supported relationships
+    std::set<core::Relationship> relationships;
+    relationships.insert(Success);
+    setSupportedRelationships(relationships);
+}
+
+void ManipulateArchive::onTrigger(core::ProcessContext *context, core::ProcessSession *session) {
+    std::shared_ptr<core::FlowFile> flowFile = session->get();
+
+    if (!flowFile) {
+        return;
+    }
+
+    std::string operation;
+    context->getProperty(Operation.getName(), operation);
+
+    std::string targetEntry;
+    context->getProperty(Target.getName(), targetEntry);
+
+    std::string destination;
+    context->getProperty(Destination.getName(), destination);
+
+    std::string before;
+    context->getProperty(Before.getName(), before);
+
+    std::string after;
+    context->getProperty(After.getName(), after);
+
+    // TODO(calebj) Validate properties
+
+    FocusArchiveEntry::ArchiveMetadata archiveMetadata;
+    FocusArchiveEntry::ReadCallback readCallback(this, &archiveMetadata);
+    session->read(flowFile, &readCallback);
+
+    logger_->log_info("ManipulateArchive performing operation %s on %s", operation.c_str(), targetEntry.c_str());
+
+    // Perform operation: REMOVE
+    if (operation == OPERATION_REMOVE) {
+        for (auto it = archiveMetadata.entryMetadata.begin(); it != archiveMetadata.entryMetadata.end();) {
+            if ((*it).entryName == targetEntry) {
+                logger_->log_info("ManipulateArchive found entry %s for removal", targetEntry.c_str());
+                std::remove((*it).tmpFileName.c_str());
```

Comment:
In the event that volatile repositories are used in environments without persistent storage should we disable this suite of extensions? As I understand it, these operations can be performed in memory, right @minifirocks?

Should we put make an option to exclude this and other processors for those without persistent storage?

Reply:
I'm leaning towards yes. There's no surefire way to determine the uncompressed size of input files.